# Deep MNIST for Experts 

*A tutorial from https://www.tensorflow.org/get_started/mnist/pros*

Loading MNIST data,

In [2]:
from tensorflow.examples.tutorials.mnist import input_data
mnist = input_data.read_data_sets('MNIST_data_pro', one_hot=True)

Extracting MNIST_data_pro/train-images-idx3-ubyte.gz
Extracting MNIST_data_pro/train-labels-idx1-ubyte.gz
Extracting MNIST_data_pro/t10k-images-idx3-ubyte.gz
Extracting MNIST_data_pro/t10k-labels-idx1-ubyte.gz


We are starting a InteractiveSession

In [3]:
import tensorflow as tf
sess = tf.InteractiveSession()

We should make a graph to understand the whole system interaction. See [Computation Graph](https://www.tensorflow.org/get_started/graph_viz). We start building the computation graph by creating nodes for the input images and target output classes.

In [4]:
x = tf.placeholder(tf.float32, shape=[None, 784])
y_ = tf.placeholder(tf.float32, shape=[None, 10])

We now define the weights W and biases b for our model,

In [5]:
W = tf.Variable(tf.zeros([784,10]))
b = tf.Variable(tf.zeros([10]))

Before **Variables** can be used within a session, they must be initialized using that session. 

In [6]:
sess.run(tf.global_variables_initializer())

We can now implement our regression model

In [7]:
y = tf.matmul(x,W) + b

Here, our loss function is the cross-entropy between the target and the softmax activation function applied to the model's prediction

In [8]:
cross_entropy = tf.reduce_mean(
    tf.nn.softmax_cross_entropy_with_logits(labels=y_, logits=y))


Now that we have defined our model and training loss function, it is straightforward to train using TensorFlow, ;) See [optimizers](https://www.tensorflow.org/api_guides/python/train#optimizers)

In [9]:
train_step = tf.train.GradientDescentOptimizer(0.5).minimize(cross_entropy)


The returned operation **train_step**, when run, will apply the gradient descent updates to the parameters. Training the model can therefore be accomplished by repeatedly running **train_step**.

In [10]:
for _ in range(1000):
  batch = mnist.train.next_batch(100)
  train_step.run(feed_dict={x: batch[0], y_: batch[1]})


Evaluate the Model

In [11]:
correct_prediction = tf.equal(tf.argmax(y,1), tf.argmax(y_,1))
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
print(accuracy.eval(feed_dict={x: mnist.test.images, y_: mnist.test.labels}))


0.9186


Getting 92% accuracy on MNIST is bad. It's almost embarrassingly bad! :(

### Build a Multilayer Convolutional Network

Now we are talking! Around 92% is sucks! So, we are trying to improve that value with a CNN. 